### PR TITLE
[RHPAM-2684] Use same pattern as for other resources

### PR DESCRIPTION
**Pull request by ruromero**
Signed-off-by: Ruben Romero Montes <rromerom@redhat.com>

Depends on https://github.com/RHsyseng/operator-utils/pull/50

### CODE REVIEW

**Comment by ruromero** [comment]
/assign @bmozaffa 
It is still using my branch for operator-utils

File: cmd/manager/main.go [hunk 1]
```diff
@@ -44,7 +44,6 @@ var (
 var log = logs.GetLogger("cmd")
 
 func printVersion() {
-	log.Info(fmt.Sprintf("Operator Version: %s", version.Version))
```
Comment (ruromero) [inline cmd/manager/main.go]: This line is the same as L50

File: cmd/ui/main.go [hunk 2]
```diff
@@ -15,7 +15,6 @@ import (
 var log = logs.GetLogger("cmd")
 
 func printVersion() {
-	log.Info(fmt.Sprintf("Operator Version: %s", version.Version))
```
Comment (ruromero) [inline cmd/ui/main.go]: This line is a duplicate of L21

File: deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml [hunk 3]
```diff
@@ -104,6 +104,12 @@ spec:
         path: consoleHost
         x-descriptors:
         - urn:alm:descriptor:org.w3:link
+      - description: The reference to the generated NamespacedDashboard ConsoleLink
```
Comment (ruromero) [inline deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml]: I regenerated CSV because I forgot to add ConsoleLink to the status

File: go.mod [hunk 4]
```diff
@@ -36,7 +36,7 @@ require (
 replace (
 	// Pin RHsyseng library versions
 	github.com/RHsyseng/console-cr-form => github.com/RHsyseng/console-cr-form v0.0.0-20200129200812-dc119cb0bd4d
-	github.com/RHsyseng/operator-utils => github.com/RHsyseng/operator-utils v0.0.0-20200310065217-8584fd29cea3
+	github.com/RHsyseng/operator-utils => github.com/ruromero/operator-utils v0.0.0-20200318153723-2b9619179901
```
Comment (ruromero) [inline go.mod]: Update again before merging

File: pkg/controller/kieapp/kieapp_controller.go [hunk 5]
```diff
@@ -80,6 +80,9 @@ func (reconciler *Reconciler) Reconcile(request reconcile.Request) (reconcile.Re
 		reconciler.setFailedStatus(instance, api.UnknownReason, err)
 		return reconcile.Result{}, err
 	}
+	if instance.GetDeletionTimestamp() != nil {
```
Comment (ruromero) [inline pkg/controller/kieapp/kieapp_controller.go]: If the resource is being deleted just don't reconcile. I don't think it is necessary, otherwise, resources deleted by the Finalizers might be re-created.

Comment (tchughesiv) [reply pkg/controller/kieapp/kieapp_controller.go]: did you test this in a cluster?

File: pkg/controller/kieapp/kieapp_controller.go [hunk 6]
```diff
@@ -149,6 +154,10 @@ func (reconciler *Reconciler) Reconcile(request reconcile.Request) (reconcile.Re
 		if err != nil {
 			return reconcile.Result{}, err
 		}
+		if resourceType == reflect.TypeOf(consolev1.ConsoleLink{}) && len(delta.Added) == 1 {
+			instance.Status.ConsoleLink = delta.Added[0].GetName()
```
Comment (bmozaffa) [inline pkg/controller/kieapp/kieapp_controller.go]: This looks really ugly! Trying to understand why we need it!

File: pkg/controller/kieapp/kieapp_controller.go [hunk 7]
```diff
@@ -555,58 +551,16 @@ func (reconciler *Reconciler) setConsoleHost(cr *api.KieApp, env api.Environment
...
-}
-
 func (reconciler *Reconciler) getKubernetesResources(cr *api.KieApp, env api.Environment) []resource.KubernetesResource {
 	var resources []resource.KubernetesResource
 	objects := filterOmittedObjects(getCustomObjects(env))
 	for _, obj := range objects {
 		resources = append(resources, reconciler.getCustomObjectResources(obj, cr)...)
 	}
-	return resources
+	return reconciler.addConsoleLinkResources(resources, cr)
```
Comment (bmozaffa) [inline pkg/controller/kieapp/kieapp_controller.go]: Change the function to a `get`, and return `nil` if one is not to be added. Do the append here. Passing on the array makes code hard to follow and error-prone.

File: pkg/controller/kieapp/kieapp_controller.go [hunk 8]
```diff
@@ -947,6 +901,16 @@ func (reconciler *Reconciler) getDeployedResources(instance *api.KieApp) (map[re
...
 	resourceMap[reflect.TypeOf(corev1.Secret{})] = secrets
 
+	if instance.Status.ConsoleLink != "" {
+		consoleLink := &consolev1.ConsoleLink{}
+		err := reconciler.Service.Get(context.TODO(), types.NamespacedName{Name: instance.Status.ConsoleLink}, consoleLink)
+		if err != nil && !errors.IsNotFound(err) {
+			log.Warn("Failed to load ConsoleLink", err)
+			return nil, err
+		}
+		resourceMap[reflect.TypeOf(*consoleLink)] = []resource.KubernetesResource{consoleLink}
```
Comment (bmozaffa) [inline pkg/controller/kieapp/kieapp_controller.go]: If `IsNotFound`, you would not want to add the empty object to the array.

File: pkg/controller/kieapp/kieapp_controller.go [hunk 9]
```diff
@@ -157,14 +166,14 @@ func (reconciler *Reconciler) Reconcile(request reconcile.Request) (reconcile.Re
...
 		}
+
 		hasUpdates = hasUpdates || added || updated || removed
 	}
 	if hasUpdates && status.SetProvisioning(instance) {
 		return reconciler.UpdateObj(instance)
 	} else if !hasUpdates {
 		log.Debug("Reconcile loop did not find any resource changes to apply")
 	}
-
```
Comment (bmozaffa) [inline pkg/controller/kieapp/kieapp_controller.go]: Why is this empty line bad, but two empty lines needed to be added in 160 and 169? :-D

File: pkg/controller/kieapp/kieapp_controller.go [hunk 10]
```diff
@@ -966,6 +930,36 @@ func (reconciler *Reconciler) getCSV(operator *appsv1.Deployment) *operatorsv1al
...
+			NamespaceDashboard: &consolev1.NamespaceDashboardSpec{
+				Namespaces: []string{cr.Namespace},
+			},
+		},
+	}
+	if cr.Status.ConsoleLink != "" {
+		consoleLink.ObjectMeta.Name = cr.Status.ConsoleLink
+	}
+	consoleLink.ObjectMeta.GenerateName = fmt.Sprintf("%s-%s-", cr.Namespace, cr.Name)
+	controllerutil.AddFinalizer(cr, constants.ConsoleLinkFinalizer)
```
Comment (bmozaffa) [inline pkg/controller/kieapp/kieapp_controller.go]: If required, should be deleted in line 165

File: pkg/controller/kieapp/kieapp_controller.go (hunk 5 above)
Comment (bmozaffa) [reply pkg/controller/kieapp/kieapp_controller.go]: I don't think we need this here

File: deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml (hunk 3 above)
Comment (bmozaffa) [reply deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml]: Also waiting to understand why ConsoleLink is needed in CR Status

Comment (bmozaffa) [reply deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml]: Thinking you need something stable that won't change or go away. The k8s resource ID of BC should give you that.

Comment (ruromero) [reply deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml]: Do you mean the KieApp UID? I think I could use that instead, yes.

File: pkg/controller/kieapp/kieapp_controller.go (hunk 6 above)
Comment (ruromero) [reply pkg/controller/kieapp/kieapp_controller.go]: I needed a way to set the generated name to the consoleLink field in the status. If I try the KieApp UID that shouldn't be necessary.
I don't like it either, too many magic values but I have been trying to find a better way without success.

File: pkg/controller/kieapp/kieapp_controller.go (hunk 9 above)
Comment (ruromero) [reply pkg/controller/kieapp/kieapp_controller.go]: Reverting, it was just myself moving code along to make up my mind. Didn't notice

File: pkg/controller/kieapp/kieapp_controller.go (hunk 10 above)
Comment (ruromero) [reply pkg/controller/kieapp/kieapp_controller.go]: the finalizer is deleted by the extendedReconciler

**Comment by ruromero** [comment]
> I'm not comfortable reviewing a PR on top of another PR that has not been reviewed. Maybe at some point, you can create a rebased branch in your repo on top of the baseline (excluding your previous PR) that we can look at with a diff?

This is a commit on top of master before merging the initial consoleLink PR
https://github.com/ruromero/kie-cloud-operator/commit/30af61f08173de1cfbb6b8ab015a988ef4ca0bcb

File: pkg/controller/kieapp/kieapp_controller.go (hunk 10 above)
Comment (bmozaffa) [reply pkg/controller/kieapp/kieapp_controller.go]: No, I'm saying that I don't see the need for a finalizer. That line 165 of the controller will delete `ConsoleLink` as needed.

Comment (ruromero) [reply pkg/controller/kieapp/kieapp_controller.go]: The finalizer string has to be added to the kieapp resource to specify that there is something that needs to be deleted.
During deletion, the extended reconciler will go through the kieapp finalizers array and look for the registered Finalizers, call the OnFinalizeFn and finally remove the finalizer string from the kieapp resource.

Comment (bmozaffa) [reply pkg/controller/kieapp/kieapp_controller.go]: You're thinking about your solution... forget about finalizers for a second and just look at the reconcile loop. It already deletes whatever shouldn't be there.

Comment (ruromero) [reply pkg/controller/kieapp/kieapp_controller.go]: Ok, I see. The fact that the comment was in this line blinded me

**Comment by ruromero** [comment]
Here you should be able to see the incremental changes before the other PR was merged https://github.com/kiegroup/kie-cloud-operator/compare/master...ruromero:linksclean?expand=1

**Comment by ruromero** [comment]
/hold
Depends on https://github.com/RHsyseng/operator-utils/pull/50

File: go.mod [hunk 11]
```diff
@@ -35,7 +36,7 @@ require (
 replace (
 	// Pin RHsyseng library versions
 	github.com/RHsyseng/console-cr-form => github.com/RHsyseng/console-cr-form v0.0.0-20200129200812-dc119cb0bd4d
-	github.com/RHsyseng/operator-utils => github.com/RHsyseng/operator-utils v0.0.0-20200318230845-8e48081eaa5a
+	github.com/RHsyseng/operator-utils => github.com/ruromero/operator-utils v0.0.0-20200324072727-841309f096a1
```
Comment (tchughesiv) [inline go.mod]: needs to be changed back to `RHsyseng` ...

File: deploy/catalog_resources/redhat/7.8/businessautomation-operator.7.8.0.clusterserviceversion.yaml [hunk 12]
```diff
@@ -211,8 +209,6 @@ spec:
                   value: registry.redhat.io/rhpam-7/rhpam-businesscentral-rhel8:7.7.0
                 - name: PAM_BC_MONITORING_IMAGE_7.7.0
                   value: registry.redhat.io/rhpam-7/rhpam-businesscentral-monitoring-rhel8:7.7.0
-                - name: PAM_PROCESS_MIGRATION_IMAGE_7.7.0
```
Comment (tchughesiv) [inline deploy/catalog_resources/redhat/7.8/businessautomation-operator.7.8.0.clusterserviceversion.yaml]: shouldn't be here... did you rebase these files as well? run `make csv` after and add to your commit?

File: pkg/controller/kieapp/kieapp_controller.go [hunk 13]
```diff
@@ -67,17 +66,35 @@ func (reconciler *Reconciler) Reconcile(request reconcile.Request) (reconcile.Re
...
 
 	// Fetch the KieApp instance
-	instance := &api.KieApp{}
+	instance := &api.KieApp{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      request.Name,
+			Namespace: request.Namespace,
+		},
+	}
+	var crDeleted bool
```
Comment (bmozaffa) [inline pkg/controller/kieapp/kieapp_controller.go]: Please revert these 5 lines, and... see comments below

File: pkg/controller/kieapp/kieapp_controller.go [hunk 14]
```diff
@@ -67,17 +66,35 @@ func (reconciler *Reconciler) Reconcile(request reconcile.Request) (reconcile.Re
...
-		reconciler.setFailedStatus(instance, api.UnknownReason, err)
+	}
+	requestedResources := []resource.KubernetesResource{}
+
+	if crDeleted {
+		deployed, err := reconciler.getDeployedResources(instance)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		_, err = reconciler.reconcileResources(instance, requestedResources, deployed)
```
Comment (bmozaffa) [inline pkg/controller/kieapp/kieapp_controller.go]: Move above 6-7 lines to if IsNotFound

File: pkg/controller/kieapp/kieapp_controller.go [hunk 15]
```diff
@@ -67,17 +66,35 @@ func (reconciler *Reconciler) Reconcile(request reconcile.Request) (reconcile.Re
...
+	var crDeleted bool
 	err := reconciler.Service.Get(context.TODO(), request.NamespacedName, instance)
 	if err != nil {
 		if errors.IsNotFound(err) {
 			// Request object not found, could have been deleted after reconcile request.
 			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
 			// Return and don't requeue
-			return reconcile.Result{}, nil
+			log.Info("Resource is being deleted. Reconcile for deletion.")
+			crDeleted = true
```
Comment (bmozaffa) [inline pkg/controller/kieapp/kieapp_controller.go]: Instead of setting this, just call reconcile and return

File: go.mod (hunk 11 above)
Comment (ruromero) [reply go.mod]: oh shoot, thanks

File: pkg/controller/kieapp/kieapp_controller.go (hunk 14 above)
Comment (bmozaffa) [reply pkg/controller/kieapp/kieapp_controller.go]: I think you might be able to just pass nil for requestedResources and it will default to empty slice... if I recall correctly

File: deploy/catalog_resources/redhat/7.8/businessautomation-operator.7.8.0.clusterserviceversion.yaml (hunk 12 above)
Comment (tchughesiv) [reply deploy/catalog_resources/redhat/7.8/businessautomation-operator.7.8.0.clusterserviceversion.yaml]: nm, i see now that this wasn't removed for 7.8.0 CSV's with prior cherry-pick... so this is expected. all good...

**Comment by tchughesiv** [comment]
@ruromero seeing similar csv changes related to PIM images in other rebased PRs... I've submitted https://github.com/kiegroup/kie-cloud-operator/pull/404 to clean that up. We'll need this PR rebased one more time once that is merged.

**Comment by ruromero** [comment]
> @ruromero seeing similar csv changes related to PIM images in other rebased PRs... I've submitted #404 to clean that up. We'll need this PR rebased one more time once that is merged.

rebased

File: deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml [hunk 16]
```diff
@@ -7,7 +7,7 @@ metadata:
     categories: Integration & Delivery
     certified: "false"
     containerImage: quay.io/kiegroup/kie-cloud-operator:7.8.0
-    createdAt: "2020-03-25 15:21:10"
+    createdAt: "2020-03-25 21:55:35"
```
Comment (bmozaffa) [inline deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml]: Should this file be reset / excluded?

File: deploy/catalog_resources/redhat/7.8/businessautomation-operator.7.8.0.clusterserviceversion.yaml [hunk 17]
```diff
@@ -7,7 +7,7 @@ metadata:
     categories: Integration & Delivery
     certified: "true"
     containerImage: registry.redhat.io/rhpam-7/rhpam-rhel8-operator:7.8.0
-    createdAt: "2020-03-25 15:21:10"
+    createdAt: "2020-03-25 21:55:35"
```
Comment (bmozaffa) [inline deploy/catalog_resources/redhat/7.8/businessautomation-operator.7.8.0.clusterserviceversion.yaml]: Should this file be reset / excluded?

File: deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml (hunk 16 above)
Comment (ruromero) [reply deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml]: done

File: deploy/catalog_resources/redhat/7.8/businessautomation-operator.7.8.0.clusterserviceversion.yaml (hunk 17 above)
Comment (ruromero) [reply deploy/catalog_resources/redhat/7.8/businessautomation-operator.7.8.0.clusterserviceversion.yaml]: done

**Comment by tchughesiv** [comment]
for reference, here's what logs look like with this PR... see this log entry a little more than I'd like, even when no KieApp has been deployed. Otherwise works as expected.
```
{"level":"info","ts":"2020-03-25T23:50:02.46719617Z","logger":"kieapp.controller","msg":"Resource is being deleted. Reconcile for deletion."}
```
https://gist.github.com/tchughesiv/476ae4a347e6cd5495696144071be6f5
